hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/GeneratedValuesProcessor.java

```diff
@@ -187,10 +191,41 @@ private JdbcParameterBindings getJdbcParameterBindings(Object id, SharedSessionC
 		return jdbcParamBindings;
 	}
 
-	private void setEntityAttributes(Object entity, Object[] state, Object[] selectionResults) {
+	private void setEntityAttributes(
+			Object entity,
+			Object[] state,
+			Object[] selectionResults,
+			EventType timing,
+			SharedSessionContractImplementor session) {
 		for ( int i = 0; i < generatedValuesToSelect.size(); i++ ) {
 			final AttributeMapping attribute = generatedValuesToSelect.get( i );
-			final Object generatedValue = selectionResults[i];
+			Object generatedValue = selectionResults[i];
+			if ( attribute.isEmbeddedAttributeMapping() ) {
+				final EmbeddableMappingTypeImpl embeddableMappingType = (EmbeddableMappingTypeImpl) attribute.getMappedType();
+				final Object previousState = state[attribute.getStateArrayPosition()];
+				if ( generatedValue == null && previousState != null ) {
+					generatedValue = embeddableMappingType.getRepresentationStrategy()
+							.getInstantiator()
+							.instantiate(
+									() -> embeddableMappingType.getValues( previousState ),
+									session.getSessionFactory()
+							);
+				}
+				else if ( previousState != null ) {
+					final AttributeMappingsList attributeMappings = embeddableMappingType.getAttributeMappings();
+					for ( int j = 0; j < attributeMappings.size(); j++ ) {
+						final AttributeMapping embeddedAttribute = attributeMappings.get( j );
+						if ( embeddedAttribute.getGenerator() == null || !embeddedAttribute.getGenerator().getEventTypes().contains( timing ) ) {
+							final PropertyAccess propertyAccess = embeddedAttribute.getAttributeMetadata()
+									.getPropertyAccess();
+							propertyAccess.getSetter().set(
+									generatedValue,
+									propertyAccess.getGetter().get( previousState )
+							);
```

Comment:
I think this may cause problems with immutable embeddable types, I would recommend using the`org.hibernate.type.CompositeType#replacePropertyValues` with a pre-populated array of values instead.